native/core/src/execution/fair_memory_pool.rs

```diff
@@ -0,0 +1,159 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{
+    fmt::{Debug, Formatter, Result as FmtResult},
+    sync::Arc,
+};
+
+use jni::objects::GlobalRef;
+
+use crate::{
+    errors::CometResult,
+    jvm_bridge::{jni_call, JVMClasses},
+};
+use datafusion::{
+    common::{resources_datafusion_err, DataFusionError},
+    execution::memory_pool::{MemoryPool, MemoryReservation},
+};
+use datafusion_execution::memory_pool::MemoryConsumer;
+use parking_lot::Mutex;
+
+/// A DataFusion fair `MemoryPool` implementation for Comet. Internally this is
+/// implemented via delegating calls to [`crate::jvm_bridge::CometTaskMemoryManager`].
+pub struct CometFairMemoryPool {
+    task_memory_manager_handle: Arc<GlobalRef>,
+    pool_size: usize,
```

Comment:
maybe this should be AtomicUsize.... this is interesting why FairSpillPool in DF is `usize` for `used` and other pools like Greedy or Unbounded are using Atomic variants. Is it related somehow to Fairpool specifics?

Reply:
`pool_size` does not need to be `AtomicUsize` because it is not meant to be updated. The number comes from a static config

Reply:
`used` is different, it gets accessed by the multiple memory consumers

Reply:
Oh, you are right I was confused between `used` and `pool_size`